File: dotnet/Lexy/Lexy.Compiler/Language/Tables/TableRow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;

namespace Lexy.Compiler.Language.Tables
{
    public class TableRow : Node
    {
        public IList<Expression> Values { get; }

        private TableRow(Expression[] values, SourceReference reference) : base(reference)
        {
            Values = values ?? throw new ArgumentNullException(nameof(values));
        }

        public static TableRow Parse(IParserContext context)
        {
            var index = 0;
            var validator = context.ValidateTokens<TableRow>();

            if (!validator.Type<TableSeparatorToken>(index).IsValid)
            {
                return null;
            }

            var tokens = new List<Expression>();
            var currentLineTokens = context.CurrentLine.Tokens;
            while (++index < currentLineTokens.Length)
            {
                var valid = !validator
                    .IsLiteralToken(index)
                    .Type<TableSeparatorToken>(index + 1)
                    .IsValid;

                if (valid)
                {
                    return null;
                }

                var reference = context.TokenReference(index);
                var token = currentLineTokens.Token<Token>(index++);
                var expression = ExpressionFactory.Parse(context.SourceCode.File, new TokenList(new[] { token }),
                    context.CurrentLine);
                if (context.Failed(expression, reference)) return null;

                tokens.Add(expression.Result);
            }

            return new TableRow(tokens.ToArray(), context.LineStartReference());
        }

        public override IEnumerable<INode> GetChildren()
        {
            return Values.ToList();
        }

        protected override void Validate(IValidationContext context)
        {
        }
    }
}

File: typescipt/src/language/tables/tableRow.ts


export class TableRow extends Node {
   public Array<Expression> Values

   private TableRow(Expression[] values, SourceReference reference) {
     super(reference);
     Values = values ?? throw new Error(nameof(values));
   }

   public static parse(context: IParseLineContext): TableRow {
     let index = 0;
     let validator = context.ValidateTokens<TableRow>();

     if (!validator.Type<TableSeparatorToken>(index).IsValid) return null;

     let tokens = new Array<Expression>();
     let currentLineTokens = context.line.tokens;
     while (++index < currentLineTokens.length) {
       let valid = !validator
         .IsLiteralToken(index)
         .Type<TableSeparatorToken>(index + 1)
         .IsValid;

       if (valid) return null;

       let reference = context.line.TokenReference(index);
       let token = currentLineTokens.Token<Token>(index++);
       let expression = ExpressionFactory.parse(new TokenList(new[] { token }), context.line);
       if (context.failed(expression, reference)) return null;

       tokens.Add(expression.result);
     }

     return new TableRow(tokens.ToArray(), context.line.lineStartReference());
   }

   public override getChildren(): Array<INode> {
     return Values.ToList();
   }

   protected override validate(context: IValidationContext): void {
   }
}
